packages/xod-project/src/patch.js

```diff
@@ -1303,17 +1304,35 @@ export const isAbstractPatch = def(
 export const validateAbstractPatch = def(
   'validateAbstractPatch :: Patch -> Either Error Patch',
   R.ifElse(
-    isAbstractPatch,
+    isAbstractPatch, // TODO: also validate composite patches!
     patch => {
-      const genericPinTypes = R.compose(
-        R.sort(R.ascend(R.identity)),
-        R.uniq,
-        R.map(Pin.getPinType),
+      const [genericInputTypes, genericOutputTypes] = R.compose(
+        R.map(R.map(Pin.getPinType)),
+        R.partition(Pin.isInputPin),
         R.filter(Pin.isGenericPin),
         listPins
       )(patch);
 
-      if (R.isEmpty(genericPinTypes)) {
+      // for example, there could be no output-t2 is there is no input-t2
+      const orphanGenericOutputTypes = R.compose(
+        R.sort(R.ascend(R.identity)),
+        R.uniq,
+        R.without
+      )(genericInputTypes, genericOutputTypes);
+
+      if (!R.isEmpty(orphanGenericOutputTypes)) {
+        return Either.Left(
+          new Error(orphanGenericOutputs(orphanGenericOutputTypes))
+        );
+      }
+
+      const allGenericPinTypes = R.compose(
+        R.sort(R.ascend(R.identity)),
+        R.uniq,
```

Comment:
I’ve found at least three places in this PR with such composition. Maybe, it makes sense to encapsulate the func in a utility:

```js
const sortUniq = R.compose(
  R.sort(R.ascend(R.identity)),
  R.uniq
);
```

Reply:
I've managed to find only two, both in `validateAbstractPatch`.
Do you think a local utility is ok?
I thinks `sortUniq` is too specific to put in `xod-func-tools`

Reply:
Ah, two of them were the same one spread over two commits. My bad. The local utility is OK.